### PR TITLE
Request for comment: creating a user service

### PR DESCRIPTION
**Pull request by drsnyder**
This service will provide a facade in front of the database layer and provide a container within which to place the user creation logic so it can be moved out of the registration controller and shared between the
standard user creation and the facebook user creation controllers.

This was something that @nmonterroso, @ArturKlajnerok, and I talked about today regarding the need for an additional abstraction for user creation/lookup.

/cc @Wikia/services-team

### CODE REVIEW

File: domain/facebookuser.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package domain
+
+type FacebookUser interface {
```
Comment (nmonterroso) [inline domain/facebookuser.go]: should `FacebookUserValue` implement this interface?

**Comment by nmonterroso** [comment]
This is great, though `UserService` probably needs a few more functions - `usernameExists(name string)`, `emailExists(name string)`, `fbIdExists(fbId int64)` and probably the other functions that are currently exposed by `UserStorage`.

File: domain/facebookuser.go (hunk 1 above)
Comment (drsnyder) [inline domain/facebookuser.go]: Yeah, I hadn't gotten that far by the time I had to stop yesterday. I wanted to share the prototype so we could get started on it today.

**Comment by drsnyder** [comment]
@Wikia/services-team  I think this is ready for review. I can't quite replace the user creation logic in the controllers until I port over the external user validation that validates users via MW and phalanx.

**Comment by ArturKlajnerok** [comment]
@drsnyder Look solid to me, I have only one question about `FacebookUser` as we won't need to create any specific user for facebook and only mapping of wikia user id and facebook user id.

File: service/user_test.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+type UserServiceTestSuite struct {
+	suite.Suite
+	userDAO   *MockedUserDAO
+	validator *MockedValidator
+	user      *domain.UserValue
+}
+
+func (suite *UserServiceTestSuite) SetupTest() {
+	suite.userDAO = new(MockedUserDAO)
+	suite.user = domain.NewUser("bill", "password", "bill@wikia-inc.com", "1990-05-01")
```
Comment (ArturKlajnerok) [inline service/user_test.go]: Values are repeated, we could move them to const.

**Comment by drsnyder** [comment]
> I have only one question about FacebookUser as we won't need to create any specific user for facebook and only mapping of wikia user id and facebook user id.

@ArturKlajnerok ok, let's talk about how this association should be represented within the domain. Something like this might make sense:

```go
facebookService.LinkUser(domain.FacebookUser)
```

What this would do is link the facebookuser id to the domain user using the username.

File: service/user_test.go (hunk 2 above)
Comment (drsnyder) [inline service/user_test.go]: Ok, I moved these to constants in 2864644.

**Comment by nmonterroso** [comment]
LGTM

**Pull request merged**